_posts/2022-01-10-reassign-partitions.md

```diff
@@ -0,0 +1,315 @@
+---
+layout: post
+title:  "Reassign partitions in Kafka Cluster"
+date: 2021-09-23
+author: shubham_rawat
+---
+
+Scaling down Kafka Brokers and increasing the number of topics and partitions are all pretty common tasks for Kafka users. Let's take a case, for eg. There are certain brokers in a cluster and now we want to remove a broker from the cluster. We need to make sure that the broker which is going to be removed should not have any assigned partitions. Now we require a tool which can assign the partitions from the broker to be scaled down, to the brokers which will be responsible for handling these partitions. The most convenient tool for this job is the reassignment partition tool: `kafka-reassign-partitions.sh`.
+
+<!--more-->
+
+## Partition Reassignment tool
+
+Within a broker pod, the `bin/kafka-reassign-partitions.sh` tool allows you to reassign partitions to different brokers.
+
+It has three different modes:
+
+| Mode | Description |
+| :-----------: | ------------- |
+| `generate`  | Takes a set of topics and brokers and generates a reassignment JSON file which will result in the partitions of those topics being assigned to those brokers. Because this operates on whole topics, it cannot be used when you only want to reassign some partitions for some topics.|
+| `execute` | Takes a reassignment JSON file and applies it to the partitions and brokers in the cluster. The `reassignment.json` file can be either the one proposed by the `--generate` command or written by the user itself. |
+| `verify`  | Using the same reassignment JSON file as the `--execute` step, `--verify` checks whether all the partitions in the file have been moved to their intended brokers. If the reassignment is complete, `--verify` also removes any traffic throttles (`--throttle`) that are in effect. Unless removed, throttles will continue to affect the cluster even after the reassignment has finished. |
+
+While using the tool, you have to provide it with two essential JSON files: `topics.json` and `reassignment.json`.
+Wondering what these two JSON files really do? Let's talk a bit about them:
+
+- The `topic.json` basically consists of the topics that we want to reassign or move. Based on this JSON file, our tool will generate a proposal `reassignment.json` which we can further change as per our requirements or use directly.
+  
+- The `reassignment.json` file is a configuration file which is used during the partition reassignment process. The reassignment partition tool will give us a proposal `reassignment.json` file when we feed it with the `topics.json` file. We can change the `reassignment.json` file as per our requirement and use it.
+
+## Why use the Partition Reassignment tool?
```

Comment:
This is the section I'd like to see at the start of the post, since it motivates the reader into why they might want to learn about this tool at all.